### ISSUE TITLE
boxed_local: False positive for using Box in C FFI

### ISSUE THREAD

**Issue by CryZe** [rust-lang/rust-clippy#5542]
In C FFI you often use Box to pass Rust types to C and back. However it seems like clippy's `boxed_local` lint is accidentally linting against this:
```rust
#[no_mangle]
pub extern "C" fn drop_the_rust_type(x: Box<String>) { }
```
```
warning: local variable doesn't need to be boxed here
 --> src/lib.rs:2:38
  |
2 | pub extern "C" fn drop_the_rust_type(x: Box<String>) { }
  |                                      ^
  |
  = note: `#[warn(clippy::boxed_local)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#boxed_local
```
[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=814b428fd192b9a9d323d0a688402d5d)

clippy 0.0.212 (204bb9b5 2020-03-17)

**Comment by olamiko** [comment]
Hi, I will like to work on this issue. But I am new to rust and clippy, I will be back with questions

**Comment by olamiko** [comment]
I found the definition of clippy's `boxed_local` in `clippy_lints`. But I am stuck on how to edit the implementation such that it doesn't throw an error when passing rust types to C (and back)

**Comment by flip1995** [comment]
You have to look at the [`FnHeader`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.FnHeader.html) which you can get from the [`FnKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/intravisit/enum.FnKind.html), which you get a a param to [`check_fn`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/passes/trait.LateLintPass.html#method.check_fn)

**Comment by olamiko** [comment]
Thank you @flip1995, I am working on it now.
